[WIP] Add Tapas model (#8113)
huggingface/transformers: docs/source/model_doc/tapas.rst

> TAPAS

Reviewer: Amazing doc!!!

and WIKITQ, but with a simpler model architecture. We additionally find that transfer learning, which is trivial in our
setting, from WIKISQL to WIKITQ, yields 48.7 accuracy, 4.2 points above the state-of-the-art.*

In addition, the authors have further pre-trained TAPAS to recognize table entailment, by creating a balanced dataset
of millions of automatically created training examples which are learned in an intermediate step prior to fine-tuning.
The authors of TAPAS call this further pre-training intermediate pre-training (since TAPAS is first pre-trained on MLM,
and then on another dataset). They found that intermediate pre-training further improves performance on SQA, achieving
a new state-of-the-art as well as state-of-the-art on TabFact, a large-scale dataset with 16k Wikipedia tables for
table entailment (a binary classification task). For more details, see their follow-up paper: `Understanding tables with
intermediate pre-training <https://arxiv.org/abs/2010.00571>`__ by Julian Martin Eisenschlos, Syrine Krichene and
Thomas Müller.

The original code can be found `here <https://github.com/google-research/tapas>`__.

> Tips:

Reviewer: Awesome that you guys added tips!

  with a causal language modeling (CLM) objective are better in that regard.

.. _original Github repository: https://huggingface.co/models?search=tapas

Usage: fine-tuning
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Here we explain how you can fine-tune :class:`~transformers.TapasForQuestionAnswering` on your own dataset. 

===========================================================================
STEP 1: Choose one of the 3 ways in which you can use TAPAS - or experiment
===========================================================================

Basically, there are 2 options: either you start from an already-finetuned checkpoint, or you start with a pre-trained base 
model and randomly initialized classification heads. In case your dataset is rather small (hundreds of training examples), 
it's advised to start from an already fine-tuned checkpoint. Both pre-trained and fine-tuned checkpoints are available in the
HuggingFace `model hub`_.

.. _model hub: https://huggingface.co/models?search=tapas

> There are 3 different ways in which one can fine-tune an already fine-tuned :class:`~transformers.TapasForQuestionAnswering` 

Reviewer: do we need the `an already fine-tuned` phrase here?

Reviewer: yeah it seems we need it! Never mind!

  You don't need this column in case you're going for WTQ/WikiSQL/WikiSQL-supervised.
- ``question``: string
- ``table_file``: string, name of a csv file containing the tabular data
- ``answer_coordinates``: list of one or more tuples (each tuple being a cell coordinate, i.e. row, column pair that is part of the answer)
- ``answer_text``: list of one or more strings (each string being a cell value that is part of the answer)
- ``aggregation_label``: index of the aggregation operator. Only required in case of strong supervision for aggregation (the WikiSQL-supervised case)
- ``float_answer``: the float answer to the question, if there is one (np.nan if there isn't). Only required in case of weak supervision for aggregation (such as WTQ and WikiSQL)

The tables themselves should be present in a folder, each table being a separate csv file. Note that the authors of the TAPAS algorithm used conversion 
scripts with some automated logic to convert the other datasets (WTQ and WikiSQL) into the SQA format. The author explains this `here`_. This is 

> actually interesting, because these conversion scripts are not perfect (the ``answer_coordinates`` and ``float_answer`` fields are populated based on the ``answer_text``), 

Reviewer: ```suggestion
Interestingly, these conversion scripts are not perfect ( ....) meaning that ...
actually interesting, because these conversion scripts are not perfect (the ``answer_coordinates`` and ``float_answer`` fields are populated based on the ``answer_text``), 
```

meaning that WTQ and WikiSQL results could actually be improved.

.. _here: https://github.com/google-research/tapas/issues/50#issuecomment-705465960

==========================================================================================
STEP 3: Convert your data into PyTorch tensors using :class:`~transformers.TapasTokenizer`
==========================================================================================

Third, given that you've prepared your data in this TSV/CSV format (and corresponding CSV files containing the tabular data), you can then 

> use :class:`~transformers.TapasTokenizer` to convert table-question pairs into input_ids, attention_mask, token_type_ids and so on. Again, 

Reviewer: ```suggestion
use :class:`~transformers.TapasTokenizer` to convert table-question pairs into :obj:`input_ids`, :obj:`attention_mask`, :obj:`token_type_ids` and so on. Again, 
```

> based on which of the three cases you picked above, :class:`~transformers.TapasForQuestionAnswering` needs different things to be fine-tuned:

Reviewer: ```suggestion
based on which of the three cases you picked above, :class:`~transformers.TapasForQuestionAnswering` requires different inputs to be fine-tuned:
```

+------------------------------------+----------------------------------------------------------------------------------------------+

:class:`~transformers.TapasTokenizer` creates the ``label_ids``, ``numeric_values`` and ``numeric_values_scale`` based on the 
``answer_coordinates`` and ``answer_text`` columns of the TSV file. The ``float_answer`` and ``aggregation_labels`` are already in the TSV file of step 2. 
Here's an example:

.. code-block::

        >>> from transformers import TapasTokenizer
        >>> import pandas as pd

        >>> model_name = 'tapas-base-finetuned-wtq'
        >>> tokenizer = TapasTokenizer.from_pretrained(model_name)

        >>> data = {'Actors': ["Brad Pitt", "Leonardo Di Caprio", "George Clooney"], 'Number of movies': ["87", "53", "69"]}
        >>> queries = ["What is the name of the first actor?", "How many movies has George Clooney played in?", "What is the total number of movies?"]
        >>> answer_coordinates = [[(0, 0)], [(1, 0)], [(0, 2), (1, 2), (2, 2)]]
        >>> answer_text = [["Brad Pitt"], ["69"], ["209"]]
        >>> table = pd.Dataframe(data)

>         >>> inputs = tokenizer(table=table, queries=queries, answer_coordinates=answer_coordinates, answer_text=answer_text, return_tensors='pt')

Reviewer: I like the API

        >>> inputs
        {'input_ids': tensor([[ ... ]]), 'attention_mask': tensor([[...]]), 'token_type_ids': tensor([[[...]]]),
        'numeric_values': tensor([[ ... ]]), 'numeric_values_scale: tensor([[ ... ]]), label_ids: tensor([[ ... ]])}

Note that :class:`~transformers.TapasTokenizer` expects the data of the table to be text-only. You can use ``.astype(str)`` on a dataframe to turn it into

> text-only data. Of course, this only shows how to encode a single training example. In reality, you should create a PyTorch dataset and a corresponding dataloader:

Reviewer: ```suggestion
text-only data. Of course, this only shows how to encode a single training example. It is advised to create a PyTorch dataset and a corresponding dataloader:
```

        >>> from transformers import TapasForQuestionAnswering

        >>> model = TapasForQuestionAnswering.from_pretrained("tapas-base-finetuned-wtq", return_dict=True)

        >>> for epoch in range(2):  # loop over the dataset multiple times
        ...    for idx, batch in enumerate(train_dataloader):
        ...         # get the inputs; 
        ...         input_ids, attention_mask, token_type_ids, label_ids, numeric_values, numeric_values_scale, float_answer = batch

        ...         # zero the parameter gradients
        ...         optimizer.zero_grad()

        ...         # forward + backward + optimize
        ...         outputs = model(input_ids=input_ids, attention_mask=attention_mask, token_type_ids=token_type_ids, 
        ...                        label_ids=label_ids, numeric_values=numeric_values, numeric_values_scale=numeric_values_scale, 
        ...                        float_answer=float_answer)
        ...         loss = outputs.loss
        ...         loss.backward()
        ...         optimizer.step()

Usage: inference
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Here we explain how you can perform inference (i.e. making predictions on new data) with :class:`~transformers.TapasForQuestionAnswering`.

Reviewer: ```suggestion
Here we explain how you can use :class:`~transformers.TapasForQuestionAnswering` for inference (i.e. making predictions on new data)
```

> Basically, for inference, you only need to provide ``input_ids``, ``attention_mask`` and ``token_type_ids`` (which you can obtain using 

Reviewer: For inference, only .... have to be provided to the model to obtain output logits.

:class:`~transformers.TapasTokenizer`) to the model to obtain the logits. Next, you can use the handy ``convert_logits_to_predictions`` method

> of :class:`~transformers.TapasTokenizer` to convert these into actual predicted coordinates and aggregation indices. 

Reviewer: ```suggestion
of :class:`~transformers.TapasTokenizer` to convert these into predicted coordinates and aggregation indices. 
```

However, note that inference is **different** depending on whether or not the setup is conversational. In a non-conversational set-up, inference 
can be done in parallel on all table-question pairs of a batch. Here's an example of that:

.. code-block::

        >>> from transformers import TapasTokenizer, TapasForQuestionAnswering
        >>> import pandas as pd 

        >>> model_name = 'tapas-base-finetuned-wtq'
        >>> model = TapasForQuestionAnswering.from_pretrained(model_name)
        >>> tokenizer = TapasTokenizer.from_pretrained(model_name)

        >>> data = {'Actors': ["Brad Pitt", "Leonardo Di Caprio", "George Clooney"], 'Number of movies': ["87", "53", "69"]}
        >>> queries = ["What is the name of the first actor?", "How many movies has George Clooney played in?", "What is the total number of movies?"]
        >>> table = pd.Dataframe(data)

>         >>> inputs = tokenizer(table, queries, return_tensors='pt')

Reviewer: no need for keyword arguments here?

Author: Fixed this, should indeed have keyword arguments + padding up to max length.

>         >>> logits, logits_agg = model(**inputs)

Reviewer: this looks like it would break given that `use_cache=True` by default now

Author: You mean `return_dict=True`?

Author: Correct? @patrickvonplaten
